Add kubevirt cert rotate doc (#73)
harvester/harvesterhci.io: kb/2024-11-28/kubevirt_cert_rotation.md

---
title: KubeVirt Certificates Rotation
description: KubeVirt Certificates Rotation.
slug: kubevirt_certificates_rotation
authors:
  - name: Cooper Tseng
    title: Software Engineer
    url: https://github.com/brandboat
    image_url: https://github.com/brandboat.png
tags: [harvester, kubevirt, certificates, cert, ca]
hide_table_of_contents: false
---

- Harvester: [Issue 5798](https://github.com/harvester/harvester/issues/5798)

> In Harvester's embedded Rancher UI, you may see some warnings on the secrets page indicating that the KubeVirt certicates are about to expire. While there is no need to worry about it as KubeVirt handles automatic certificate rotation. 

Reviewer: There are also logs from related PODs, better to add some examples.

# KubeVirt Certificate Rotation Strategy
KubeVirt applies self-signed certificate mechamism, where both CA and certifcates are rotated on a defined recurring interval. You can check the corresponding setting  `certificateRotateStrategy` by using cmd
```sh
kubectl get kubevirt -n harvester-system -o yaml
```
By default, the `certificateRotateStrategy` is empty, meaning KubeVirt will use its default rotation settings, and no manual configuration is needed.
```yaml
certificateRotateStrategy: {}
```
A complete configuration look like this:
```yaml
certificateRotateStrategy:
  selfSigned:
    ca:
      duration: 24h
      renewBefore: 4h
    server:
      duration: 168h
      renewBefore: 28h
```
Explanation of Fields:
- `.ca.duration`: refers to the validity period the CA certificate. The default value is 7 days.
- `.ca.renewBefore`: refers to how much time before the CA certificate expiration a new certificate should be issued. The default value is 7 days * 0.2.
- `.server.duration`: refers to the validity period of server certificates (e.g. virt-api, virt-handler, virt-operator, etc.). The default value is 1 day.

> - `.server.renewBefore`: refers to how much time before the server certificates expire a new certificate should be issued. The default value is 1 day * 0.2.

Reviewer 2: How are the default values actually specified? I'm asking because the config example you provided uses hours while the default values all use "day" as the unit of measurement.

> - `.server.renewBefore`: refers to how much time before the server certificates expire a new certificate should be issued. The default value is 1 day * 0.2.

Author: My bad, the default values are
- `.ca.duration` : 168h
- `.ca.renewBefore`: 33.6h
- `.server.duration`: 24h
- `.server.renewBefore`: 4.8h

And thanks for the suggestion, the unit should all be `h` here. Behind the scene KubeVirt use [Duration](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/duration) as unit.

I'll update everything in the latest commit, thanks.

## Triggers for Certificate Rotation

> The following conditions can trigger certificate rotation in KubeVirt, note that this list highlights key triggers and is not exhaustive.

Reviewer: when kubevirt rotates the certs, are there a short time-window ? that will cause new VMs are failed to be created?

Reviewer 3: Good point. From what I can see, there is an overlap period where the old and new certs are merged and served, before the old one expired: https://github.com/kubevirt/kubevirt/blob/ec514a29d5552cf5442f262a140681c099e124c8/pkg/virt-operator/resource/apply/core.go#L533-L551

Author: Thank you both !
> when kubevirt rotates the certs, are there a short time-window ? that will cause new VMs are failed to be created?

I tried to test this by setting KubeVirt CR `certificateRotateStrategy` to
```yaml
certificateRotateStrategy:
  selfSigned:
    ca:
      duration: 60s
      renewBefore: 10s
    server:
      duration: 30s
      renewBefore: 5s
configuration:
```

Then I tried to pause/unpause vm multiple times, and it turns out during the cert rotation, there might have a chance that KubeVirt may be unable to handle requests.
![image](https://github.com/user-attachments/assets/bce171cf-56dc-4737-8c32-ed784c36ad2d)

Author: Perhaps we need to warn users about [the comment above](https://github.com/harvester/harvesterhci.io/pull/73#discussion_r1872811080). I’d appreciate your thoughts on this. Thanks! c.c. @bk201

Reviewer: When the side effect is trully existing, we may consider:
(1) Add this formally in the document;
(2) Evaluate a proper default value on Harvester, is current KubeVirt's default value fitting for Harvester? Expose it to Harvester setting so to let user have an option to adjust.
(3) Have a place on the UI to show how long it will expire, user can monitor it, trigger the rotation actively e.g. on a cluster maintainence period.

Reviewer 4: @w13915984028 The PR should be used to clarify the current kubevirt cert rotate behavior, for any improvement it's out of scope and please file an issue to track.

> ## Triggers for Certificate Rotation
> The following conditions can trigger certificate rotation in KubeVirt, note that this list highlights key triggers and is not exhaustive.

Reviewer 2: ```suggestion
## Certificate Rotation Triggers

Certificate rotation can be triggered by several conditions. The following list only outlines key triggers and is not exhaustive.
```

> - Missing Certificates: If a certificate doesn’t exist.
> - Invalid CA Signature: If a certificate wasn’t signed by the specified CA.
> - Proactive Renewal: Certificate is within the `renewBefore` period before expiration.
> - CA Expiration: If the CA certificate expires, the certificate signed by the CA will also be rotated.

Reviewer 2: ```suggestion
- Missing certificate: A required certificate does not exist.
- Invalid CA signature: A certificate was not signed by the specified CA.
- Proactive renewal: The `renewBefore` value takes effect. A new certificate must be issued before the current one expires.
- CA expiration: The CA certificate has expired, so the certificate signed by the CA is also rotated.
```

> ## References
> - https://kubevirt.io/2020/KubeVirt-Security-Fundamentals.html
> - https://github.com/kubevirt/kubevirt/blob/v1.1.1/pkg/virt-operator/resource/generate/components/secrets.go#L326
> - https://github.com/kubevirt/kubevirt/blob/v1.1.1/pkg/virt-operator/resource/apply/certificates.go

Reviewer 2: ```suggestion
## References

- Related GitHub issue: [#5798](https://github.com/harvester/harvester/issues/5798)
- https://kubevirt.io/2020/KubeVirt-Security-Fundamentals.html
- https://github.com/kubevirt/kubevirt/blob/v1.1.1/pkg/virt-operator/resource/generate/components/secrets.go#L326
- https://github.com/kubevirt/kubevirt/blob/v1.1.1/pkg/virt-operator/resource/apply/certificates.go
```

```sh
kubectl get kubevirt -n harvester-system -o yaml
```
By default, the value of `certificateRotateStrategy` is empty, which means that KubeVirt uses its default rotation settings and no manual configuration is required.
```yaml
certificateRotateStrategy: {}
```

## Configuration Fields

You can use the following fields to configure `certificateRotateStrategy`.
- `.ca.duration`: Validity period of the CA certificate. The default value is 168 hours.
- `.ca.renewBefore`: Amount of time before a CA certificate expires during which a new certificate is issued. The default value is 33.6 hours
- `.server.duration`: Validity period of server component certificates (for example, virt-api, virt-handler, and virt-operator). The default value is 24 hours.
- `.server.renewBefore`: Amount of time before a server certificate expires during which a new certificate is issued. The default value is 4.8 hours.

Example of a complete configuration:
```yaml
certificateRotateStrategy:
  selfSigned:
    ca:
      duration: 168h
      renewBefore: 33.6h
    server:

>       duration: 24d

Author: Good catch

      renewBefore: 4.8h
```

## Certificate Rotation Triggers

Certificate rotation can be triggered by several conditions. The following list only outlines key triggers and is not exhaustive.
- Missing certificate: A required certificate does not exist.
- Invalid CA signature: A certificate was not signed by the specified CA.
- Proactive renewal: The `renewBefore` value takes effect. A new certificate must be issued before the current one expires.
- CA expiration: The CA certificate has expired, so the certificate signed by the CA is also rotated.

> When certificate rotation is triggered, you should see logs in deployment `virt-operator` as belows:

Reviewer 2: ```suggestion
When certificate rotation is triggered, you should see `virt-operator` log records similar to the following:
```

Author: Thank you ! I have addressed all comments in the latest commit.
